File: testdata/fixtures/clojure/CWE-78.clj
;; CWE-78: OS Command Injection – user input passed to shell/sh
;; allows arbitrary command execution. Safe code uses literals.

(ns fixture.cwe78
  (:require [clojure.java.shell :refer [sh]]))

(defn vuln [request]
  (let [cmd (get-in request [:params :cmd])]
    ;; ruleid: clojure-command-injection
    (sh cmd))
  (let [arg (get-in request [:params :arg])]
    ;; ruleid: clojure-command-injection
    (clojure.java.shell/sh arg))
  (let [body (get-in request [:params :body])]
    ;; ruleid: clojure-command-injection
    (sh "echo" body)))

(defn safe []
  ;; ok: clojure-command-injection
  (sh "ls" "-la")
  ;; ok: clojure-command-injection
  (clojure.java.shell/sh "echo" "hi")
  ;; ok: clojure-command-injection
  (sh "echo" "fixed"))

File: testdata/fixtures/elixir/CWE-78.ex
# CWE-78: OS Command Injection – user input passed to System.cmd/shell
# allows arbitrary command execution. Literals are safe.

defmodule Fixture.CWE78 do
  def vuln(user_input) do
    # ruleid: elixir-command-injection
    System.cmd(user_input, ["-la"])
    # ruleid: elixir-command-injection
    System.shell(user_input)
    # ruleid: elixir-command-injection
    System.cmd(user_input, ["arg"])
    # ruleid: elixir-command-injection
    cmd = user_input; System.shell(cmd)
  end

  def safe do
    # ok: elixir-command-injection
    System.cmd("ls", ["-la"])
    # ok: elixir-command-injection
    System.shell("echo hi")
    # ok: elixir-command-injection
    System.cmd("echo", ["fixed"])
  end
end
